src/storage/v3/splitter.cpp

```diff
@@ -0,0 +1,240 @@
+  data.vertices = CollectVertices(data, collected_transactions_, split_key);
+  data.edges = CollectEdges(collected_transactions_, data.vertices, split_key);
+  data.transactions = CollectTransactions(collected_transactions_, data.vertices, *data.edges);
+
+  return data;
+}
+
+void Splitter::ScanDeltas(std::set<uint64_t> &collected_transactions_, Delta *delta) {
+  while (delta != nullptr) {
+    collected_transactions_.insert(delta->commit_info->start_or_commit_timestamp.logical_id);
+    delta = delta->next;
+  }
+}
+
+VertexContainer Splitter::CollectVertices(SplitData &data, std::set<uint64_t> &collected_transactions_,
+                                          const PrimaryKey &split_key) {
+  // Collection of indices is here since it heavily depends on vertices
+  // Old vertex pointer new entry pointer
+  std::map<LabelId, std::multimap<const Vertex *, const LabelIndex::IndexContainer::iterator>>
+      label_index_vertex_entry_map;
+  std::map<std::pair<LabelId, PropertyId>,
+           std::multimap<const Vertex *, const LabelPropertyIndex::IndexContainer::iterator>>
+      label_property_vertex_entry_map;
+
+  data.label_indices =
+      CollectIndexEntries<LabelIndex, LabelId>(indices_.label_index, split_key, label_index_vertex_entry_map);
+  data.label_property_indices = CollectIndexEntries<LabelPropertyIndex, std::pair<LabelId, PropertyId>>(
+      indices_.label_property_index, split_key, label_property_vertex_entry_map);
+  // This is needed to replace old vertex pointers in index entries with new ones
+  const auto update_indices = [](auto &entry_vertex_map, auto &updating_index, const auto *old_vertex_ptr,
+                                 auto &new_vertex_ptr) {
+    for ([[maybe_unused]] auto &[index_type, vertex_entry_mappings] : entry_vertex_map) {
+      auto [it, end] = vertex_entry_mappings.equal_range(old_vertex_ptr);
+      while (it != end) {
+        auto entry_to_update = *it->second;
+        entry_to_update.vertex = &*new_vertex_ptr;
+        updating_index.at(index_type).erase(it->second);
+        updating_index.at(index_type).insert(std::move(entry_to_update));
+        ++it;
+      }
+    }
+  };
+
+  VertexContainer splitted_data;
+  auto split_key_it = vertices_.find(split_key);
+  while (split_key_it != vertices_.end()) {
+    // Go through deltas and pick up transactions start_id/commit_id
+    ScanDeltas(collected_transactions_, split_key_it->second.delta);
+
+    const auto *old_vertex_ptr = &*split_key_it;
+    auto next_it = std::next(split_key_it);
+
+    const auto &[splitted_vertex_it, inserted, node] = splitted_data.insert(vertices_.extract(split_key_it->first));
+    MG_ASSERT(inserted, "Failed to extract vertex!");
+
+    // Update indices
+    update_indices(label_index_vertex_entry_map, data.label_indices, old_vertex_ptr, splitted_vertex_it);
+    update_indices(label_property_vertex_entry_map, data.label_property_indices, old_vertex_ptr, splitted_vertex_it);
```

Comment:
Does the pointer to vertex changes? I think extract + insert shouldn't move the underlying pair in the memory at all. However I might misunderstood why is this necessary.

Reply:
You are right the address does not change, I assumed it would. Yes I thought this is needed to adjust the vertex ptr in entries, but it seems redundant, and I will remove it and test